test/modules/EdgeHubRestartTester/src/MessageEdgeHubConnector.cs

```diff
@@ -0,0 +1,111 @@
+                Settings.Current.MessageOutputEndpoint,
+                this.runExpirationTime,
+                this.cancellationToken).ConfigureAwait(false);
+
+            TestResultBase msgTestResult = new EdgeHubRestartMessageResult(
+                Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartMessage.ToString(),
+                DateTime.UtcNow,
+                Settings.Current.TrackingId,
+                this.batchId.ToString(),
+                Interlocked.Read(ref this.messageCount).ToString(),
+                this.edgeHubRestartedTime,
+                msgCompletedTime,
+                mgsStatusCode,
+                this.restartSequenceNumber);
+
+            var reportClient = new TestResultReportingClient { BaseUrl = Settings.Current.ReportingEndpointUrl.AbsoluteUri };
+            await ModuleUtil.ReportTestResultUntilSuccessAsync(
+                reportClient,
+                this.logger,
+                msgTestResult,
+                this.cancellationToken).ConfigureAwait(false);
+        }
+
+        async Task<Tuple<DateTime, HttpStatusCode>> SendMessageAsync(
+            ModuleClient moduleClient,
+            string trackingId,
+            Guid batchId,
+            string msgOutputEndpoint,
+            DateTime runExpirationTime,
+            CancellationToken cancellationToken)
+        {
+            while ((!cancellationToken.IsCancellationRequested) && (DateTime.UtcNow < runExpirationTime))
+            {
+                Message message = new Message(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(new { data = DateTime.UtcNow.ToString() })));
+                Interlocked.Increment(ref this.messageCount);
+                message.Properties.Add("sequenceNumber", Interlocked.Read(ref this.messageCount).ToString());
+                message.Properties.Add("batchId", batchId.ToString());
+                message.Properties.Add("trackingId", trackingId);
+
+                try
+                {
+                    // Sending the result via edgeHub
+                    await moduleClient.SendEventAsync(msgOutputEndpoint, message);
+                    this.logger.LogInformation($"[SendMessageAsync] Send Message with count {Interlocked.Read(ref this.messageCount).ToString()}: finished.");
+                    return new Tuple<DateTime, HttpStatusCode>(DateTime.UtcNow, HttpStatusCode.OK);
+                }
+                catch (TimeoutException ex)
+                {
+                    // TimeoutException is expected to happen while the EdgeHub is down.
+                    // Let's log the attempt and retry the message send until successful
+                    this.logger.LogDebug(ex, $"[SendMessageAsync] Exception caught with SequenceNumber {this.messageCount}, BatchId: {batchId.ToString()};");
+                    // Make sure the seqeunce number is not increment when the sent failed.
+                    Interlocked.Decrement(ref this.messageCount);
```

Comment:
Increment messageCount before while loop and don't need to decrease it.

Reply:
messageCount is always reported as 1 with current implementation.

Reply:
Why do you need locking on this.messageCount?  Only 1 thread is running for this code, right?

Reply:
It may be running on 1 thread. But since `SendMessageAsync()` is called with `ConfigureAwait(false);` there is no guarantee that it is running on the same synchronization context. Hence, I'm being intentional here to lock it.